### ISSUE TITLE
Configuring Ranger

### ISSUE THREAD

**Issue by mfix22** [badges/shields#3931]
:clipboard: **Description**

I noticed that you installed [Ranger](https://reporanger.com) but have not yet used it to automate merging PRs, closing issues, or handling other tasks. 

Have you had any trouble with it, or need any help configuring it? I am here to help! 👍 

If not, feel free to close this issue (or have Ranger do it for you 🙂)

**Comment by calebcartwright** [comment]
hi @mfix22, thanks for reaching out. We did try to use it a while back on our PRs but were unable to get it to work. I still think we'd really benefit from the functionality, but we didn't really have time to dig into why it wasn't working/what we may have been missing/etc.

If you have any insight on how we could get it working that would be great!

One example of a PR where we tried: https://github.com/badges/shields/pull/3526

**Comment by mfix22** [comment]
Hey @calebcartwright sorry to hear that! I would love to help make sure it is working correctly for you. I see that all the checks passed on your PR, so that is looking good ✅ Did you have any other protections on the `master` branch at all? 

On a fresh clone of the repo, I see that Ranger is working fine, https://github.com/mfix22/shields/pull/1#event-2592619178, so I am assuming it has something to do with your other integrations, most likely CI/CD. 

If you could try again with a small/dummy PR, that would be great, otherwise I will try and set up CircleCI on my fork to better recreate the situation.

**Comment by chris48s** [comment]
We use quite a lot of branch protection rules on this repo. My assumption is that one of these rules is preventing repo ranger from being able to merge for us. We have the following rules enabled:

* Require pull request reviews before merging
* Required approving reviews: 1
* Dismiss stale pull request approvals when new commits are pushed
* Require status checks to pass before merging
* Require branches to be up to date before merging
* Include administrators
* Restrict who can push to matching branches: Organization and repository administrators, badges/shields-core

Any thoughts?

While its nice that you've proactively contacted us about this, one of the reasons we've not followed this up at all is that its not very obvious how to contact you with a support query. Even if your code is not public, have you considered setting up a public issues-only repo or something?
Thanks

**Comment by paulmelnikow** [comment]
Just wanted to say @mfix22 that a merge queue we could kick off after an approving review that would take care of updating the branch, waiting for tests to pass (except LGTM and Docker, which take too long), and then squash-merging, would be _super helpful_ for this project. I'd estimate it would save 1–2 hours of maintainer time a week.

**Comment by mfix22** [comment]
> While it's nice that you've proactively contacted us about this, one of the reasons we've not followed this up at all is that its not very obvious how to contact you with a support query. Even if your code is not public, have you considered setting up a public issues-only repo or something?
Thanks

@chris48s this is great feedback, thank you. We have a public feedback form you can contact me at here: https://reporanger.com/feedback, however, I will also set up a public issues-only repo for tracking bugs/issues here: https://github.com/reporanger/feedback.

**Comment by mfix22** [comment]
@paulmelnikow that is great to hear — I would absolutely love to help you save those hours. 

If you add the "squash when passing" label to a PR, Ranger will automatically squash-merge the PR once all the required statuses are passing, following all integration status checks, branch protection rules, etc. 

I you can add the label to a branch in need of merging, I can investigate why it might not be being merged.

**Comment by paulmelnikow** [comment]
Can you look at the example @calebcartwright posted? https://github.com/badges/shields/pull/3526

**Comment by mfix22** [comment]
> Can you look at the example @calebcartwright posted? #3526

On it 👌

**Comment by paulmelnikow** [comment]
#3880 has the necessary approvals.

<img width="761" alt="Screen Shot 2019-08-29 at 1 34 23 PM" src="https://user-images.githubusercontent.com/1487036/63962919-d377e800-ca61-11e9-9a76-aae9993fbb50.png">

This is what we have under "restrict who can push to matching branches".

<img width="738" alt="Screen Shot 2019-08-29 at 1 34 48 PM" src="https://user-images.githubusercontent.com/1487036/63962936-e1c60400-ca61-11e9-916e-8c45d39ffdfc.png">

Do we need to add a user for Ranger?

**Comment by mfix22** [comment]
Hey @paulmelnikow I think you are right! The public status listing shows that the PR has a `mergeable_state` of `clean`, which means "good to go", however Ranger is getting the status: `blocked` which means that is is not able to merge the PR. 

Can you try adding Ranger to an approved list you mentioned above? We have never seen this specific set of branch blocking rules before so I _really_ appreciate you are letting me debug this 🙂

**Comment by paulmelnikow** [comment]
I can only add GitHub users. Which user is it?

**Comment by mfix22** [comment]
@repo-ranger is the username.

**Comment by paulmelnikow** [comment]
It's added!

<img width="745" alt="Screen Shot 2019-08-29 at 2 16 52 PM" src="https://user-images.githubusercontent.com/1487036/63965676-ac241980-ca67-11e9-9c43-562d1bb1f489.png">

**Comment by calebcartwright** [comment]
Looks like the PR was merged!

**Comment by mfix22** [comment]
Great! 👍 Thanks for working through this with me. I added another set of logging to our backend that will help me understand this issue faster in the future 👌 

Now that we got it working, are there any other features you would like help automating? On other repos I maintain, I configure things like: automatically deleting HEAD branches, deleting ":+1:" comments, stuff like that 🙂

**Comment by paulmelnikow** [comment]
Awesome!

Will it update an out-of-date branch before attempting to merge it? That's the thing we need for this to be useful.

**Comment by calebcartwright** [comment]
> Will it update an out-of-date branch before attempting to merge it?

And would we need that to be done via a dependabot command on dependabot opened PRs to avoid merge conflicts, and if so can ranger support that?

**Comment by mfix22** [comment]
We are actively building out that feature, but trying to figure out how to best fit it in with the rest of Ranger's config.

I will ping in this thread, or in a new issue immediately once the feature ships.

**Comment by chris48s** [comment]
Thanks for helping with this. Even without the ability to auto-`@dependabot rebase` this will probably save a bit of time.
It would be helpful to mention in the documentation that users need to manually give @repo-ranger write access to master in the branch protection config as well as authorising the app. I suspect we won't be the first users not to understand this.

**Comment by mfix22** [comment]
@chris48s done! reporanger.com/docs — Under the section on "Merging Pull Requests"

Also I have just pushed the feature that will keep a branch up to date as long as your branch protection rules require that. Would you like me to enable that feature for you repo?

**Comment by chris48s** [comment]
Nice. Yes. Lets give it a try. Thanks

**Comment by paulmelnikow** [comment]
@mfix22 I noticed the merge queue is waiting on LGTM to run, which is not a required check, mostly because it is sometimes very slow!

Can I request a feature that only the _required_ checks must pass in order to merge?

Or alternatively, that we can list certain checks to ignore?

**Comment by paulmelnikow** [comment]
The automatic branch updating is working!

<img width="797" alt="Screen Shot 2019-09-03 at 8 25 59 PM" src="https://user-images.githubusercontent.com/1487036/64216933-1e8e6280-ce89-11e9-9514-2a74fb1d0c6a.png">

**Comment by mfix22** [comment]
Hey @paulmelnikow 👋 looks like Ranger no longer has permission to merge your branches. Did a branch protection rule recently get updated? You might have to give Ranger write access to re-enable PR merging.  

You can find how to handle this under the "Branch Protection" section of the [Repo Ranger docs](http://reporanger.com/docs)

**Comment by calebcartwright** [comment]
Seems to have just merged one in #5026 no?

**Comment by mfix22** [comment]
Oh, okay. Great!

**Comment by paulmelnikow** [comment]
Thanks for reaching out!

We had an issue where maintainers couldn't click the "Update Branch" button. It might have been due to a setting on the PR or maybe it was some GitHub fluke. See https://github.com/badges/shields/pull/5008#issuecomment-622978456

Maybe that was related?

**Comment by mfix22** [comment]
That could be! Thanks @paulmelnikow. I'm just glad it is still working 👍
